sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -617,6 +618,77 @@ object NullPropagation extends Rule[LogicalPlan] {
 }
 
 /**
+ * Propagate foldable expressions:
+ * Replace all attributes with aliases of the original foldable expressions except the followings.
+ * 1) Command and Set(UNION/INTERSECT/EXCEPT): Do not optimize.
+ * 2) Filter/Sort: Use the original foldable expressions without aliases.
```

Comment:
instead of doing this, can't we always use alias and run `CleanupAlias` for the result?